Feature/core 498 add form data component

karladler opened the pull request.

karladler committed:
CORE-498 add action button and form data wrapper

WIP

karladler committed:
CORE-498 add action button and form data wrapper

WIP

karladler committed:
Merge branch 'develop' into FEATURE/CORE-498-add-form-data-component

karladler committed:
remove empty lines at EOF

karladler committed:
CORE-498 remove action-button and merge functionality into evan-button

karladler committed:
make buttons icon framework agnostic

Commit:
Merge branch 'develop' into FEATURE/CORE-498-add-form-data-component

Tschuck commented on dapps/evancore.vue.libs/src/components/form-data-input/form-data-input.vue:
```diff
@@ -0,0 +1,50 @@
+        @focus="$parent.$emit('setFocus')"
+        :id="id"
+        class="form-control"
+        v-bind:value="value"
```
we can use newer `@value`, `@input` notifcation

Tschuck replied:
we can use newer `@value`, `@input` notifcation

Tschuck commented on dapps/evancore.vue.libs/src/components/form-data-select/form-data-select.vue:
```diff
@@ -0,0 +1,54 @@
+      <select
+        @focus="$parent.$emit('setFocus')" class="form-control"
+        :id="id"
+        v-bind:value="value"
```
we can use newer `@value`, `@input` notifcation

Commit:
Merge branch 'FEATURE/CORE-498-add-form-data-component' of github.com:evannetwork/ui-vue into FEATURE/CORE-498-add-form-data-component

Tschuck commented on dapps/evancore.vue.libs/src/components/button/button.scss:
```diff
@@ -0,0 +1,15 @@
+.button-wrapper {
```
missing license headers

Tschuck commented on dapps/evancore.vue.libs/src/components/form-data-wrapper/form-data-wrapper.scss:
```diff
@@ -0,0 +1,27 @@
+@import '~@evan.network/ui/src/style/utils';
```
missing license headers

Tschuck reviewed (commented):
In general nothing to criticise. Contrary to my initial rejection of the components, I find the resulting template very good. :D

The only thing i missing, is the [ui-docs](https://github.com/evannetwork/ui-docs) component documentation ;)

Some enhancements proposals:

1. Can we use the form-data component even for other formular specifications? What happens in formulars with stacked labels? (e.g. Onboarding) What happens in formulars with stacked lables and grid layout? (e.g. postcode and city in one line)

2. What do you think about a general form-data-wrapper click handler that handles setFocus on input elements? So input components do not need to send the editMode events. But it could be possible in this case, that custom input will not be detected.

```js
handleInputClick($event) {
    const inputElements = [ 'input', 'select' ];

    if (inputElements.indexOf($event.target.nodeName.toLowerCase()) !== -1) {
      this.editMode = true;
    }
  }
```

3. Should we implement a general disabled state handling? Currently, the formular button will be enabled, also when the formular is invalid.
4. Would it be useful, to add `EvanFormControl` support? Currently in every application that uses a formular, we use the `EvanForm` specification, where validation and value handling's are implemented. If we add a type parameter, it would be possible to dynamically create formulars from definition. Also the disabled state handling can be handled dynamically within the `form-data-component`. Even custom inputs can be added, by iterating over each input and generation slots for them. Per default, the dynamic one can be created, else it can also be overwritten by developer.

```js
new EvanForm(this, {
  alias: {
    value: '',
    type: 'string',
    validate: function(vueInstance: ContactAddComponent, form: ContactFormInterface) {
      return this.value.length !== 0;
    }
  },
  tags: {
    value: '',
    type: 'string',
  },
  emailInvite: {
    value: false,
    validate: function(vueInstance: ContactAddComponent, form: ContactFormInterface) {
      // trigger the validation process for the address and email contorl,
      // when the email invite was changed
      form.accountId.validate();
      form.email.validate();

      return true;
    }
  },
  select: {
    value: '',
    type: 'string',
    options: [
      { label: '1', value: '1' },
      { label: '2', value: '2' },
      { label: '3', value: '3' }
    ],
  }
}, '_addressbook.contact-form');
```
5. Current save handler uses one promise for displaying loading state. Should we add a loading param? Or should we add dispatcher support for handling loading states in data-form-wrapper? So the component can trigger the save function, which will usually starts a dispatcher instance, and also watch for dispatcher finalisation, to remove the loading state. Dispatcher logic also animates the user to use other applications while the data is saving. After reentering the component, the click save handler would not take effect.

Commit:
add click event to profile-preview

- [CORE-426]

Commit:
Merge branch 'FEATURE/CORE-498-add-form-data-component' into FEATURE/CORE-426-switch-profile-type

Commit:
Merge branch 'develop' into FEATURE/CORE-426-switch-profile-type

Commit:
add form-data-wrapper evan form support

- [CORE-426]

Commit:
add blur events to form-data elements; add correct translation handling

- [CORE-498]

Commit:
merge lowercase feature branch

- [CORE-498]

Commit:
refactor data-control naming; add more properties for controls

- move form data controls to forms folder and rename to from-control
- add form control component
- add file form control support
- add sizing for controls
- [CORE-498]

Commit:
fix inline controls

- [CORE-498]

Commit:
fix files design

- [CORE-498]

Commit:
fix file control design

- [CORE-498]
